crates/hir_ty/src/diagnostics/expr.rs

```diff
@@ -40,7 +41,7 @@ impl<'a, 'b> ExprValidator<'a, 'b> {
     }
 
     pub(super) fn validate_body(&mut self, db: &dyn HirDatabase) {
-        let body = db.body(self.owner.into());
+        let (body, source_map) = db.body_with_source_map(self.owner.into());
```

Comment:
Calling `body_with_source_map` will cause this to be recomputed whenever any change even to whitespace is done in the function, so we need to be careful to only call it when necessary (i.e. when there are actual diagnostics to emit).

Reply:
I'll call it when we have type mismatches